# НИС «Основы анализа данных в Python»

*Алла Тамбовцева*

## Практикум 3. Датафреймы `pandas`: фильтрация наблюдений и введение в описательные статистики

In [1]:
import pandas as pd

Загрузите данные из файла `Stars.csv` (скачать [здесь](https://github.com/allatambov/PyDat25/blob/main/Stars.csv)) и выполните следующие задания.

Показатели в файле:

* `Temperature`: температура звезды по Кельвину;
* `Luminosity`: светимость звезды в единицах светимости Солнца;
* `Radius`: радиус звезды по отношению к Солнцу;
* `Absolute magnitude`: абсолютная звездная величина;
* `Star type`: тип звезды (0 — коричневый карлик, 1 — красный карлик, 2 — белый карлик, 3 —
звезда главной последовательности, 4 — супергигант, 5 — гипергигант);
* `Star color`: цвет звезды;
* `Spectral Class`: спектральный класс звезды.

Если в датафрейме есть строки с пропущенными значениями, удалять их не нужно.

In [2]:
stars = pd.read_csv("Stars.csv")
stars.head()

,Temperature,Luminosity,Radius,Absolute magnitude,Star type,Star color,Spectral Class
0,3068,0.002400,0.1700,16.12,0,Red,M
1,3042,0.000500,0.1542,16.60,0,Red,M
2,2600,0.000300,0.1020,18.70,0,Red,M
3,2800,0.000200,0.1600,16.65,0,Red,M
4,1939,0.000138,0.1030,20.06,0,Red,M


### Задача 0

Выведите информацию о размерности датафрейма и типах столбцов в датафрейме. Выведите описательные статистики для всех столбцов датафрейма.

Подсказка: описательные статистики все сразу – метод `.describe()` и этот же метод с аргументом `include = "object"`.

In [ ]:
print(stars.shape)
print(stars.dtypes)

(240, 7)
Temperature             int64
Luminosity            float64
Radius                float64
Absolute magnitude    float64
Star type               int64
Star color             object
Spectral Class         object
dtype: object


In [3]:
# описательные статистики для всех числовых столбцов,
# по умолчанию выбираются типы float и int

stars.describe()

,Temperature,Luminosity,Radius,Absolute magnitude,Star type
count,240.000000,240.000000,240.000000,240.000000,240.000000
mean,10497.462500,107188.361635,237.157781,4.382396,2.500000
std,9552.425037,179432.244940,517.155763,10.532512,1.711394
min,1939.000000,0.000080,0.008400,-11.920000,0.000000
25%,3344.250000,0.000865,0.102750,-6.232500,1.000000
50%,5776.000000,0.070500,0.762500,8.313000,2.500000
75%,15055.500000,198050.000000,42.750000,13.697500,4.000000
max,40000.000000,849420.000000,1948.500000,20.060000,5.000000


> Здесь `count` – число заполненных ячеек, `mean` – среднее арифметическое, `std` – стандартное отклонение, `min` и `max` – минимум и максимум, `50%` – медиана, `25%` – 25-ый процентиль, он же нижний квартиль, он же первый квартиль, он же квантиль уровня 0.25, `75%` – 75-ый процентиль, он же верхний квартиль, он же третий квартиль, он же квантиль уровня 0.75.

In [4]:
# включаем в выдачу описание только текстовых столбцов
# типа object

stars.describe(include = "object")

,Star color,Spectral Class
count,240,240
unique,19,7
top,Red,M
freq,112,111


> Здесь `count` – число заполненных ячеек, `unique` – число уникальных значений (категорий), `top` – мода, самое частое значение, `freq` – частота, соответствующая моде. Так, например, звезды по спектральному классу делятся на 7 категорий, самая частая из них – категория `M`, таких звезд в файле 111.

### Задача 1

Сохраните строки, соответствующие звездам с абсолютной звездной величиной (`Absolute magnitude`) ниже 0, в датафрейм `low`. Вычислите количество таких строк.

In [5]:
low = stars[stars["Absolute magnitude"] < 0]
print(low.shape[0], len(low)) # любой вариант

102 102


### Задача 2

Проверьте, есть ли в данных звезды, которые являются белыми карликами, чей цвет отличен от белого (значение `White`).

In [6]:
# выбираем строки
q02 = stars[(stars["Star type"] == 2) & (stars["Star color"] != "White")]

# проверяем, правда ли, что число строк не 0

print(q02.shape[0] != 0)

True


### Задача 3

Сохраните строки, соответствующие звездам спектрального класса `O` и `M` , в датафрейм `stars_om`. Вычислите среднюю температуру таких звезд и округлите ее до второго знака после запятой.

In [7]:
# вариант 1

stars_om = stars[(stars["Spectral Class"] == "O") |
                 (stars["Spectral Class"] == "M")]
print(stars_om.shape)

(151, 7)


In [8]:
# вариант 2

stars_om = stars[stars["Spectral Class"].isin(["O", "M"])]
print(stars_om.shape)

(151, 7)


In [10]:
# считаем среднюю температуру
# второй вариант – если версия Python/Numpy старовата

print(stars_om["Temperature"].mean().round(2))
print(round(stars_om["Temperature"].mean(), 2))

8300.18
8300.18


### Задача 4

Сохраните строки, соответствующие звездам-сверхгигантам и звездам-супергигантам, в датафрейм `giants`. Вычислите минимальную светимость таких звезд.

In [11]:
# вариант 1

giants = stars[(stars["Star type"] == 4) | (stars["Star type"] == 5)]
print(giants["Luminosity"].min())

74000.0


In [12]:
# вариант 2

giants = stars[stars["Star type"].isin([4, 5])]
print(giants["Luminosity"].min())

74000.0


In [13]:
# вариант 3

giants = stars[stars["Star type"] >= 4]
print(giants["Luminosity"].min())

74000.0


### Задача 5

Сохраните строки, соответствующие звездам главной последовательности синего (`Blue`) или сине-белого цвета (`Blue-white` или `Blue-White`), в датафрейм `main_blue`.

Вычислите количество строк в полученном датафрейме. Выведите описательные статистики для столбца `Spectral Class`. Звезд какого спектрального класса больше всего среди отобранных наблюдений?

In [14]:
# вариант 1: ключевой, обратите внимание
# на наличие дополнительных скобок

main_blue = stars[(stars["Star type"] == 3) &
      ((stars["Star color"] == "Blue") |
       (stars["Star color"] == "Blue-white") |
       (stars["Star color"] == "Blue-White"))]

print(main_blue.shape[0])

26


In [15]:
# вариант 2: .isin() вместо ИЛИ

main_blue = stars[(stars["Star type"] == 3) &
                  (stars["Star color"].isin(["Blue", "Blue-white",
                                             "Blue-White"]))]

print(main_blue.shape[0])

26


In [16]:
# вариант 3: если понравился str.contains(),
# пользуемся тем, что Blue-white и Blue-White
# отличаются только регистром, поэтому
# можем выключить учет регистра case = False


main_blue = stars[(stars["Star type"] == 3) &
     (stars["Star color"].str.contains("Blue") |
     stars["Star color"].str.contains("Blue-white", case = False))]
print(main_blue.shape[0])

26


In [17]:
# больше всего звезд класса B

main_blue["Spectral Class"].describe()

,Spectral Class
count,26
unique,3
top,B
freq,12


### Задача 6

Сохраните строки, соответствующие звездам, чья температура не ниже медианы (50-ый процентиль), но не выше верхнего квартиля (75-ый процентиль), в датафрейм `last`.

In [19]:
last = stars[(stars["Temperature"] >= stars["Temperature"].median()) &
     (stars["Temperature"] <= stars["Temperature"].quantile(0.75))]
print(last.shape)

(60, 7)
